### ISSUE TITLE
QUESTION: Dynamic schema classes

### ISSUE THREAD

**Issue by nikoladsp** [zircote/swagger-php#977]
Hi,
I am thinking on how to best solve following problem. Say I have a simple "model" class:

```
<?php declare(strict_types=1);

namespace API\Model;

/**
 * @OA\Schema()
 */
class User {
   /**
     * User's ID
     * @var int
     * @OA\Property()
     */
    public int $id;

    /**
     * User's first name
     * @var string
     * @OA\Property()
     */
    public string $first_name;

    /**
     * User's last name
     * @var string
     * @OA\Property()
     */
    public string $last_name;
}
```

Now, when I want to create User object - **$first_name** and **$last_name** are mandatory fields, while **$id** must not be passed in JSON payload from the client app (front-end). On the other hand when updating existing user, **$id** is mandatory while other fields are optional.

So some pseudo-code would be:

```
class Schema
{
}

class CreateUser extend Schema
{
    private User $model_class; // User from the above example
    private array $required_fields = array('first_name', 'last_name');
    private array optional_fields = array();
}

class UpdateUser extend Schema
{
    private User $model_class; // User from the above example
    private array $required_fields = array('id');
    private array optional_fields = array('first_name', 'last_name');
}
```

So my dilemma/question is how to make these schemes for CRUD operations flexible enough so I don't have to copy paste code and phpdoc for properties?

I am thinking on using [laminas](https://github.com/laminas/laminas-code) to generate classes using a script that will analyze all subclasses of Schema and based on given model and required/optional fields would simply generate PHP classes which will not be committed in GIT repository, but rather run when app starts.

Basically, I would like to use Swagger and schema validation but without having to do a bunch of copy paste code.

Is there a better way?

Thank you in advance

**Comment by DerManoMann** [comment]
You can compose schemas using `allOf`, but that doesn't give you options with regards to required/optional fields. Not sure how those get merged in that case.

**Comment by nikoladsp** [comment]
Maybe I over-complicated the question. Shortest possible version would be: how to create schema classes dynamically which contains docblocks, without using eval? Is this possible with anonymous classes?

Best regards

**Comment by DerManoMann** [comment]
Not sure why you would need eval but I might be still missing something.

If you generate the classes you can also generate the schemas, right? I would suspect the dynamic nature is just in the models not the controllers?

Either way, you can also generate schemas without annotating actual classes. Something like this also works - maybe it helps?

```
/**
 * @OA\Schema(
 *     schema="newUser",
 *     @OA\Property(property="username", type="string"),
 *     @OA\Property(property="email", type="string")
 * )
 */

/**
 * @OA\Schema(
 *     schema="userResource",
 *     allOf={
 *         @OA\Schema(
 *             @OA\Property(property="id", type="integer")
 *         )
 *         @OA\Schema(ref="#components/schemas/newUser")
 *     }
 * )
 */
```

**Comment by nikoladsp** [comment]
Thank you, I think I can work with this - would be cleaner.

Model is indeed "fixed", but I would like to generate API docs for it, based on endpoint schema "requirements". 

If I go with only defining schemes without classes, are there any validator you can recommend for it, or I shall maybe create simple one on my own?

Many thanks

**Comment by DerManoMann** [comment]
Do you mean validating the generated specs?
For specs I use https://github.com/stoplightio/spectral and https://github.com/APIDevTools/swagger-cli

**Comment by nikoladsp** [comment]
Ah sorry - I meant to validate JSON payload against specific schema. For example if username and email are required fields to check if email is valid email and for example username has minimal length of 6. Something like that.

Regards

**Comment by DerManoMann** [comment]
Ah, nothing you could use out of the box. My openapi-verifier library does this but in the context of tests that you run against your controllers.

https://github.com/DerManoMann/openapi-verifier/blob/master/src/VerifiesOpenApi.php uses the Json schema validation project to load the spec, find the matching bit of schema and then validate :)

**Comment by nikoladsp** [comment]
Thank you kindly. I will close this for now.

Many thanks and best regards
